### PR TITLE
feat(chart): add post-delete cleanup hook

### LINKED ISSUE #2931: Research and build POC for Helm post-delete hook for OSM resources
This GitHub Issue is to explore ways to run a "garbage collector" and sweep everything that belongs to OSM during `osm mesh delete` or `helm delete` so that no artifacts remain regardless of what manages these.

---

### Context
We have had conversations about what components of the OSM Kubernetes ecosystem should be installed and owned by Helm, and what should be managed by the OSM Controller.

One concern that surfaced through these debates is the remainder of ghost artifacts or components that are not cleaned after OSM is removed.

Here is a summary of the concern for situations where OSM manages a component and not Helm:
  1. User installs OSM Controller using `osm install` (or `helm install`)
  2. OSM Controller creates an object, which Helm is unaware of - let's say a ConfigMap
  3. OSM is removed from the cluster using `osm mesh delete` or `helm delete` - Helm will remove everything it installed.
  4. The ConfigMap, which OSM Controller created in step 2 will remain on the cluster (ghost object)

---

### Examples

Here is an experiment I had to test Kubernetes jobs deleting certain assets:
Available in this gist as well: https://gist.github.com/draychev/bb3d94e2379b575215a6d8b91ab9ac45
```bash
#!/bin/bash

set -aueo pipefail

NS=osm-system

kubectl apply -f - <<EOF
---
apiVersion: batch/v1
kind: Job
metadata:
  name: delete-osm-artifacts-helm-delete-job
  namespace: ${NS}
  labels:
    app: osm-garbage-collector

spec:
  template:
    metadata:
      name: delete-osm-artifacts-helm-delete-job
      namespace: ${NS}
    spec:
      serviceAccount: osm
      serviceAccountName: osm
      restartPolicy: Never
      containers:
        - name: garbage-collector
          image: bitnami/kubectl
          imagePullPolicy: IfNotPresent
          command: ['sh', '-c', 'sleep 15 && kubectl get all -A || true && kubectl delete configmap -n osm-system osm-config']
EOF

kubectl describe job -n ${NS} delete-osm-artifacts-helm-delete-job  ||  true

kubectl get jobs -n ${NS} --selector app=osm-garbage-collector  ||  true

kubectl get pods -n ${NS} --selector app=osm-garbage-collector  ||  true

kubectl logs -n ${NS} --selector job-name=delete-osm-artifacts-helm-delete-job -f  ||  true

echo -e "\n\nGet job logs:\n\n\tkubectl logs -n ${NS} --selector job-name=delete-osm-artifacts-helm-delete-job -f\n"

# How do we clean-up the Job itself?
echo -e "\n\nDelete job:\n\n\tkubectl delete job -n ${NS} --selector app=osm-garbage-collector\n"
```

--- 

### References
  - #2826  -  osm mesh upgrade reverts configmap value changes which were made out of band
  - #2386  -  helm upgrade overrides manual edits in osm-config
  - #2925  -  Label all OSM deployed Kubernetes resources with a common key/value

### PR DESCRIPTION
**Pull request by nojnhuh**
**Description**:
This change adds a Helm hook to be run after the chart resources have
been uninstalled with `helm uninstall` or `osm mesh uninstall`. The hook
is implemented as a Kubernetes Job that will run `kubectl` commands to
delete resources created outside the chart associated with the control plane.

Currently only the osm-config ConfigMap will be cleaned up by the hook
(in preparation for #2932).

Resources NOT cleaned up by the hook:
- CRDs (as they may still be in use)
- osm-ca-bundle (as it may be used for a future OSM installation)
- Envoy bootstrap secrets (because injected Pods still mount them as volumes)

Fixes #2931

**Affected area**:

- New Functionality      [X]
- Documentation          [ ]
- Install                [ ]
- Control Plane          [ ]
- CLI Tool               [ ]
- Certificate Management [ ]
- Networking             [ ]
- Metrics                [ ]
- SMI Policy             [ ]
- Security               [ ]
- Tests                  [ ]
- CI System              [ ]
- Demo                   [ ]
- Performance            [ ]
- Other                  [ ]

Please answer the following questions with yes/no.

- Does this change contain code from or inspired by another project? If so, did you notify the maintainers and provide attribution?
No

### CODE REVIEW

**Review by snehachhabria — APPROVED** [review]

**Comment by shashankram** [comment]
> * osm-ca-bundle (as it may be used for a future OSM installation)

I see the `osm-ca-bundle` secret needing to be used rarely for this purpose. Although similar to existing pods leveraging the envoy bootstrap config secret, existing ingress resources could also be referencing this to serve HTTPS to existing backends.

**Review by shashankram — APPROVED** [review]

File: charts/osm/templates/cleanup-hook.yaml [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+      restartPolicy: Never
+      containers:
+        - name: garbage-collector
+          image: bitnami/kubectl
+          imagePullPolicy: IfNotPresent
+          command:
+            - sh
+            - -c
+            - >
+             kubectl delete --ignore-not-found configmap -n '{{ include "osm.namespace" . }}' osm-config;
```
Comment (draychev) [inline charts/osm/templates/cleanup-hook.yaml:28]: @nojnhuh this is already great!

I wonder if we could expand the scope of the delete operation and point it to all remaining OSM resources. We can leverage the changes in #3068 for this.

Something like
```bash
kubectl delete all -n $NS --selector "app.kubernetes.io/name=openservicemesh"
```

Comment (nojnhuh) [reply charts/osm/templates/cleanup-hook.yaml:28]: @draychev I mentioned why I'm not sure that's the best approach (for now) here: https://github.com/openservicemesh/osm/issues/2931#issuecomment-812052480. 

tl;dr Helm already does most of the work, it's not as simple as a one-liner, and it requires much more extensive RBAC permissions.

Do you think the benefits to the all-in-one approach currently outweigh those concerns?

**Review by michelleN — COMMENTED** [review]
@nojnhuh did you look into using the `"helm.sh/hook-weight": "5"` annotation to ensure that the Role/RoleBinding/ServiceAccount were created before the Job? or did you find that Helm created those resources in the right order for you?

**Comment by nojnhuh** [comment]
> @nojnhuh did you look into using the `"helm.sh/hook-weight": "5"` annotation to ensure that the Role/RoleBinding/ServiceAccount were created before the Job? or did you find that Helm created those resources in the right order for you?

@michelleN I do remember hitting this issue and needing to add the hook weight to the Job. IIRC I had defined the Job and the RBAC resources all in the same file, and splitting the Job out solved the issue without needing to add the hook weight. Now I'm unable to recreate that same behavior, perhaps because we recently updated our Helm dependency version.

Now even if everything is in the same file, resources are created in the same order no matter the order they appear in the file.

File: charts/osm/templates/cleanup-hook.yaml (hunk 1 above)
Comment (nojnhuh) [reply charts/osm/templates/cleanup-hook.yaml:28]: @draychev I'm going to merge this for now. If we need something more generic we can readdress this then.
